### PR TITLE
feat: forum support

### PR DESCRIPTION
**Pull request by LordOfPolls**
## What type of pull request is this?

- [x] Non-breaking code change
- [ ] Breaking code change
- [ ] Documentation change/addition
- [ ] Tests change

## Description

Adds support for forum channels

## Changes

- Add required endpoint
- Create `GuildForum` object
- Create `create_forum_post` method 

## Checklist

- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

### CODE REVIEW

**Review by zevaryx — APPROVED** [review]

**Comment by Wolfhound905** [comment]
I would like to have more than one person test. So I just hope you got everything right the first time ig lol

**Comment by LordOfPolls** [comment]
On-hold due to api instability

**Comment by i0bs** [comment]
I'll test this later this week, I want to provide a consistent review that corrects some inaccuracies

**Comment by i0bs** [comment]
Is this PR meant to stay open with the merge of #469?

**Comment by Wolfhound905** [comment]
> Is this PR meant to stay open with the merge of #469?

I said yesterday. This pr (#459) will be going into naff branch. I was making a pr to the forum_support branch...

**Comment by i0bs** [comment]
Not sure why I'm getting that reaction. There's literally 4 development branches active right now! 😂 how am I supposed to know which one is the final iteration for where the PR intends to be done?

image

What I'm understanding here is that the PR was made for `forum_support` which acts as the interim between the snek/naff conversion, then there's the `naff` branch which acts as the rewrite and then... `naff_raw` exists?

**Comment by zevaryx** [comment]
`naff_raw` is a rewrite of the cache to minimize deserialization into object into an as-needed basis, instead of on receive. This should increase performance by a lot in the long run once it's finished

**Comment by LordOfPolls** [comment]
This pr is merging into the main staging branch. Wolf made a pr which was explicitly communicated to be merging into this pr/branch. 

Naff-raw is a branch zev is using for testing some raw serialisation system. 

I'm unsure what confusion you're having, but If you're really struggling DM me.

**Comment by i0bs** [comment]
The confusion I have is where new features are actually being added. Your [contribution requirements](https://github.com/Discord-Snake-Pit/Dis-Snek/blob/master/CONTRIBUTING.md#requirements) explicitly state that PRs, and in most cases, must be targeting the `dev` branch. There's three other branches, two of them admittedly are for the NAFF migration which isn't a problem since it's temporary, but then there's a `forum_support` one. I don't understand the intent of having numerous development branches.

This additionally conflicts with a past statement given on these requirements laid out in the support server.

image
> For context, [that pipeline] is a reference to the [contribution requirements of Hypothesis.](https://github.com/HypothesisWorks/hypothesis/blob/master/CONTRIBUTING.rst)

For new contributors, such as myself, this adds as a layer of confusion for how you're supposed to be PRing new features. Are we always going to get new branches now for WIP/closed beta features, or will we continue to target them on `dev`?

**Comment by LordOfPolls** [comment]
`dev` is and always has been our main development branch. As you can see it Is staging the final dis_snek release. Obviously we need to continue work on the lib while that release stages. 

The naff branch is our holding the rename and refactor of the lib as we move into becoming naff. It's not ideal to have this tempory branch but while the dis_snek release is pending, it's maditory.

New developers should, as always, target Dev. As per the contribution guidelines you highlighted. As well as the screenshot of myself. This has not changed. 

Any other branches should be treated as standone in-progress branches. Where people are working on features. This branch for example is for forums. Zevs is for serialisation. Etc

File: naff/api/http/http_requests/threads.py [hunk 1]
```diff
@@ -189,3 +190,44 @@ async def create_thread(
...
             payload["invitable"] = invitable
             return await self.request(Route("POST", f"/channels/{channel_id}/threads"), data=payload, reason=reason)
+
+    async def create_forum_thread(
+        self,
+        channel_id: "Snowflake_Type",
+        name: str,
+        auto_archive_duration: int,
+        message: dict | FormData,
+        applied_tags: List[str] = None,
```
Comment (Kigstn) [inline naff/api/http/http_requests/threads.py:201]: Suppling a list of `applied_tags` (as str) does not work - discord returns a bad payload error

Should be tested when the feature releases and the API is stable ( :) )

**Review by zevaryx — APPROVED** [review]

**Pull request merged**